draft-ietf-tls-tls13.md

```diff
@@ -1906,10 +1823,78 @@ Subsequent Handshake:
+allows server to decline resumption and fall back to a full handshake.
+A KeyShare MUST also be sent if the client is attempting to
+negotiate an (EC)DHE-PSK cipher suite.
+
+
+### Zero-RTT Data
+
+When resuming via a PSK, clients can also send data on their first
+flight ("early data"). This data is encrypted solely under keys
+derived using the PSK as the static secret.  As shown in
+{{tls-0-rtt}}, the Zero-RTT data is just added to the 1-RTT handshake
+in the first flight, the rest of the handshake uses the same messages.
+
+~~~
+         Client                                               Server
+
+         ClientHello
+           + early_data
+           + key_share*
+         (Finished)
+         (Application Data*)
+         (end_of_early_data)        -------->
+                                                         ServerHello
+                                                        + early_data
+                                                         + key_share
+                                               {EncryptedExtensions}
+                                               {CertificateRequest*}
+                                                          {Finished}
+                                   <--------     [Application Data*]
+         {Certificate*}
+         {CertificateVerify*}
+         {Finished}                -------->
+
+         [Application Data]        <------->      [Application Data]
+
+               *  Indicates optional or situation-dependent
+                  messages that are not always sent.
+
+               () Indicates messages protected using keys
+                  derived from the static secret.
+
+               {} Indicates messages protected using keys
+                  derived from the ephemeral secret.
+
+               [] Indicates messages protected using keys
+                  derived from the master secret.
+~~~
+{: #tls-0-rtt title="Message flow for a zero round trip handshake"}
+
+[[OPEN ISSUE: Should it be possible to combine 0-RTT with the
+server authenticating via a signature
+https://github.com/tlswg/tls13-spec/issues/443]]
+
+
+IMPORTANT NOTE: The security properties for 0-RTT data (regardless of
+the cipher suite) are weaker than those for other kinds of TLS data.
+Specifically:
+
+1. This data is not forward secret, because it is encrypted solely
+with the PSK.
+
+2. There are no guarantees of non-replay between connections.
+Unless the server takes special measures outside those provided by TLS (See
+{{replay-properties}}), the server has no guarantee that the same
+0-RTT data was not transmitted on multiple 0-RTT connections.
+This is especially relevant if the data is authenticated either
+with TLS client authentication or inside the application layer
+protocol. However, 0-RTT data cannot be duplicated within a connection (i.e., the server
+will not process the same data twice for the same connection) and also
+cannot be sent as if it were ordinary TLS data.
```

Comment:
What is the point of the last bit here? "also cannot be sent as if it were ordinary TLS data".

Is this a reference to API requirements?